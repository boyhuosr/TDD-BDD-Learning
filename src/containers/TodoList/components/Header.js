import React, { Component } from 'react';

class Header extends Component {
    constructor(props) {
        super(props)
        this.handleInputChange = this.handleInputChange.bind(this)
        this.hanldeInputKeyUp = this.hanldeInputKeyUp.bind(this)
        this.state={
            value: ''
        }
    }

    handleInputChange(e){
        this.setState({
            value: e.target.value
        })
    }

    hanldeInputKeyUp(e){
        const {value} = this.state
        if(e.keyCode === 13 && value) {
            this.props.addUndoItem(value)
        }
    }

    render() {
        const {value} = this.state
        return (
            <div>
                <input data-test='input' 
                    value={value} 
                    onChange={this.handleInputChange}
                    onKeyUp = {this.hanldeInputKeyUp}
                />
            </div>
        )
    }
}

export default Header;
