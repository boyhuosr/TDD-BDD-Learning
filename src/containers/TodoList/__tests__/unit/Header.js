import React from 'react';
import Header from '../../components/Header';
import {shallow} from 'enzyme';
import {findTestWrapper} from '../../../../utils/testUtils'
 

describe("Header component", ()=>{
  it('render success', ()=>{
    const wrapper = shallow(<Header />)
    expect(wrapper).toMatchSnapshot() 
  }) 
  it('Header component has a input area', ()=>{
    const wrapper = shallow(<Header />)
    //const inputElem = wrapper.find("[data-test='input']")
    const inputElem = findTestWrapper(wrapper, 'input')
    expect(inputElem.length).toBe(1);
  }) 

  it('the initial content of input should be null', ()=>{
    const wrapper = shallow(<Header />)
    const inputElem = findTestWrapper(wrapper, 'input')
    expect(inputElem.prop('value')).toEqual('');
  }) 

  it('the content of input should be changed based on input', ()=>{
    const wrapper = shallow(<Header />)
    const inputElem = findTestWrapper(wrapper, 'input')
    const userInput = 'learn Jest today'
    inputElem.simulate('change', {
        target: {value: userInput}
    })
    expect(wrapper.state('value')).toEqual(userInput);

    // const newInputElem = wrapper.find("[data-test='input']")
    // expect(newInputElem.prop('value')).toBe(userInput)
  }) 

  it('if click enter, with no content in input, no response', ()=>{
      const fn  = jest.fn()
      const wrapper = shallow(<Header addUndoItem={fn}/>)
      const inputElem = findTestWrapper(wrapper, 'input')
      wrapper.setState({value: ''})
      inputElem.simulate('keyup', {keyCode: 13})
      expect(fn).not.toHaveBeenCalled();
  })

  it('if click enter, with content in input, fn should be recalled', ()=>{
    const fn  = jest.fn()
    const wrapper = shallow(<Header addUndoItem={fn}/>)
    const inputElem = findTestWrapper(wrapper, 'input')
    const userInput = 'learn React'
    wrapper.setState({value: userInput})
    inputElem.simulate('keyUp', {keyCode: 13})
    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenLastCalledWith(userInput)

  })
})






  