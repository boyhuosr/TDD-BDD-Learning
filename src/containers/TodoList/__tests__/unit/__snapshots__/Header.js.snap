// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component render success 1`] = `
<div
  className="header"
>
  <div
    className="header-content"
  >
    TodoList
    <input
      className="header-input"
      data-test="input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="To do"
      value=""
    />
  </div>
</div>
`;
